### ISSUE TITLE
Rule proposal: Enforce dictionary type style

### ISSUE THREAD

**Issue by ark120202** [typescript-eslint/typescript-eslint#795]
Currently there are few ways to declare a dictionary-like type:

```ts
// Key can be only `string` and `number`
type _ = { [key: string]: string };
// Key is any type usable as object key
type _ = { [P in string]: string };
// Same as above
type _ = Record<string, string>;
```

I think it might make sense to implement it as 2 rules:

- Disallow object types (or interfaces) with a single index signature member and fix it to mapped type, or enforce the opposite:  disallow mapped types of `string` or `number` with unused key.
- Disallow mapped types with unused key and fix it to `Record`. It also may be extended to enforcing   usage of other builtin types, like `Partial` or `Readonly`. Not sure if there is a need to do the opposite, since it can be done with ban-types.

**Comment by sindresorhus** [comment]
I would like to enforce using `{[key: string]: string};` instead of `Record<string, string>;`. Is that part of what you've described above? It's not clear to me whether it is.

Also keep in mind that if you're going to enforce `Record<string, string>;` over `{[key: string]: string};`, you need to add an exception for `{[key: string]: any};` when it's a parameter, as it has special behavior:

> The index signature `{[key: string]: any}` in TypeScript behaves specially: it’s a valid assignment target for any object type. This is a special rule, since types with index signatures don’t normally produce this behavior.

**Comment by bradzacher** [comment]
> I would like to enforce using `{[key: string]: string};` instead of `Record<string, string>;`

I'm curious - why? Just you prefer the style, or something more?
To me the `Record` version is easier to parse and understand.

> Also keep in mind that if you're going to enforce Record<string, string>; over {[key: string]: string};, you need to add an exception for {[key: string]: any}; when it's a parameter, as it has special behavior

have you got an example where `{[key: string]: any}` works, but `Record<string, any>` doesn't? I believe they are always equivalent?

Unless I'm mistaken, you should actually be able to do the even safer
`{[key: string]: unknown}` / `Record<string, unknown>` and have any object assign to it.

**Comment by sindresorhus** [comment]
> I'm curious - why? Just you prefer the style, or something more?

Consistency, since I thought I had to use exactly `{[key: string]: any}` for input parameters, and then I would want to use the same syntax (`{[key: string]: unknown}`) for other things.

> have you got an example where {[key: string]: any} works, but Record<string, any> doesn't? I believe they are always equivalent?

No, I was just under the assumption that only exactly `{[key: string]: any}` worked. If `Record<string, any>` works too that would be great.

You can see how I had to change all my TypeScript ambient files: https://github.com/sindresorhus/camelcase-keys/commit/f276ecfa02f2a3368810bd1dd04fab99da7a7540

> Unless I'm mistaken, you should actually be able to do the even safer
{[key: string]: unknown} / Record<string, unknown> and have any object assign to it.

https://devblogs.microsoft.com/typescript/announcing-typescript-3-5-rc/#-k-string-unknown--is-no-longer-a-wildcard-assignment-target

**Comment by bradzacher** [comment]
> I was just under the assumption that only exactly {[key: string]: any} worked. If Record<string, any> works too that would be great.

Nope, the record approach works. AFAIk TS doesn't have hardcoded signatures, it just requires the types resolve to the "special thing".

```ts
declare function fnWildcard(arg: { [k: string]: any }): void;
declare function fnWildcardRecord(arg: Record<string, any>): void;
declare function someFunc(): void;

fnWildcard(someFunc);
fnWildcardRecord(someFunc);
```
[repl](https://www.typescriptlang.org/play/?ssl=6&ssc=1&pln=5&pc=22#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVNQHUsJgxZgAKWAcwC54BveAbQGsGBnDGLVGgLoMoqAJ7wAvgEoGANxxZgAbgBQoSLAQp02PAWKlyMYACVwOY9Rj14ZsBeAAebr34AaeCNEA+GfHmKqurQcEhomLj4nDgAtiAAYuGUfgHKKiqIRCRkFJTRcYnoUqqZBjnGdg55sQnhxSpAA)

---

Again, I'm curious, why would you use types that implicitly treat functions and objects the same?
Surely the intention of your `camelcase-keys` util is to only allow objects?

Not judging, etc - just wanting to understand the usecase fully so I can make an informed decision.

**Comment by sindresorhus** [comment]
> Huh. I've never needed/wanted a wildcard assignment, so I didn't realise this was a thing.

It's more like users expect that to work and then complain to me when it doesn't. I got so many issues with complaints when I switched from `{[key: string]: any}` to `{[key: string]: unknown}` in my projects.

> Again, I'm curious, why would you use types that implicitly treat functions and objects the same?
Surely the intention of your camelcase-keys util is to only allow objects?

I wouldn't.

These are my options:

1. `{ [k: string]: unknown }` - Users complain they cannot pass in object references. (Example: https://github.com/sindresorhus/camelcase-keys/pull/46)
2. `{ [k: string]: any }` - It also allows functions.

I'm all ears if you have a better way to deal with situations like https://github.com/sindresorhus/camelcase-keys/pull/46?
